source/server/admin/stats_request.cc

```diff
@@ -0,0 +1,146 @@
+#include "source/server/admin/stats_request.h"
+
+namespace Envoy {
+namespace Server {
+
+StatsRequest::StatsRequest(Stats::Store& stats, bool used_only, bool json,
+                           Utility::HistogramBucketsMode histogram_buckets_mode,
+                           absl::optional<std::regex> regex)
+    : used_only_(used_only), json_(json), histogram_buckets_mode_(histogram_buckets_mode),
+      regex_(regex), stats_(stats) {}
+
+Http::Code StatsRequest::start(Http::ResponseHeaderMap& response_headers) {
+  if (json_) {
+    render_ =
+        std::make_unique<StatsJsonRender>(response_headers, response_, histogram_buckets_mode_);
+  } else {
+    render_ = std::make_unique<StatsTextRender>(histogram_buckets_mode_);
+  }
+
+  // Populate the top-level scopes and the stats underneath any scopes with an empty name.
+  // We will have to de-dup, but we can do that after sorting.
+  //
+  // First capture all the scopes and hold onto them with a SharedPtr so they
+  // can't be deleted after the initial iteration.
+  stats_.forEachScope(
+      [this](size_t s) { scopes_.reserve(s); },
+      [this](const Stats::Scope& scope) { scopes_.emplace_back(scope.getConstShared()); });
+
+  startPhase();
+  return Http::Code::OK;
+}
+
+bool StatsRequest::nextChunk(Buffer::Instance& response) {
+  if (response_.length() > 0) {
+    ASSERT(response.length() == 0);
+    response.move(response_);
+    ASSERT(response_.length() == 0);
+  }
+  while (!render_->isChunkFull(response)) {
+    while (stat_map_.empty()) {
+      switch (phase_) {
+      case Phase::TextReadouts:
+        phase_ = Phase::CountersAndGauges;
+        startPhase();
+        break;
+      case Phase::CountersAndGauges:
+        phase_ = Phase::Histograms;
+        startPhase();
+        break;
+      case Phase::Histograms:
+        render_->render(response);
+        return false;
+      }
+    }
+
+    auto iter = stat_map_.begin();
+    const std::string& name = iter->first;
+    StatOrScopes& variant = iter->second;
+    switch (variant.index()) {
+    case 0:
+      populateStatsForCurrentPhase(absl::get<ScopeVec>(variant));
+      break;
+    case 1:
+      renderStat<Stats::TextReadoutSharedPtr>(name, response, variant);
+      break;
+    case 2:
+      renderStat<Stats::CounterSharedPtr>(name, response, variant);
+      break;
+    case 3:
+      renderStat<Stats::GaugeSharedPtr>(name, response, variant);
+      break;
+    case 4: {
+      auto histogram = absl::get<Stats::HistogramSharedPtr>(variant);
+      auto parent_histogram = dynamic_cast<Stats::ParentHistogram*>(histogram.get());
+      if (parent_histogram != nullptr) {
+        render_->generate(response, name, *parent_histogram);
+      }
+    }
+    }
+    stat_map_.erase(iter);
+  }
+  return true;
+}
```

Comment:
The role of `phase` is to ensure that different types of stats can be processed in order. But is it necessary? 
`TextReadout` is hardly ever used. 
`Counter` and `Gauge` are handled in the same batch in your implementation.
`Histogram` are always processed last.

So, why can't we just deal with each scope one by one. For stats of type `TextReadout`, `Couter` or `Gauge` are added directly to response. 
For `Histogram` stats, they are cached for final processing. 

Is there any particular reason to do it like this (your current implmentation)?

Reply:
Yes and I will add comments. This is to avoid having to change the expected results in tests. I don't want to completely change the implementation and the tests in one PR. But we can contemplate a change to co-mingle the types and reduce the number of passes.